vm_insnhelper.c

```diff
@@ -3187,6 +3186,105 @@ vm_find_or_create_class_by_id(ID id,
     }
 }
 
+static void
+vm_search_method_wrap(
+    const struct rb_control_frame_struct *reg_cfp,
+    struct rb_call_info *ci,
+    struct rb_call_cache *cc,
+    VALUE recv)
+{
+    vm_search_method(ci, cc, recv);
+}
+
+static void
+vm_search_invokeblock(
+    const struct rb_control_frame_struct *reg_cfp,
+    struct rb_call_info *ci,
+    struct rb_call_cache *cc,
+    VALUE recv)
+{
+    /* Does nothing. */
+}
+
+static VALUE
+vm_invokeblock_i(
+    struct rb_execution_context_struct *ec,
+    struct rb_control_frame_struct *reg_cfp,
+    struct rb_calling_info *calling,
+    const struct rb_call_info *ci,
+    struct rb_call_cache *cc)
+{
+    VALUE block_handler = VM_CF_BLOCK_HANDLER(GET_CFP());
+
+    if (block_handler == VM_BLOCK_HANDLER_NONE) {
+	rb_vm_localjump_error("no block given (yield)", Qnil, 0);
+    }
+    else {
+        return vm_invoke_block(ec, GET_CFP(), calling, ci, block_handler);
+    }
+}
+
+MJIT_FUNC_EXPORTED VALUE vm_exec(rb_execution_context_t *, int);
+
+static VALUE
+vm_sendish(
+    struct rb_execution_context_struct *ec,
+    struct rb_control_frame_struct *reg_cfp,
+    struct rb_call_info *ci,
+    struct rb_call_cache *cc,
+    VALUE block_handler,
+    void (*explorer)(
+        const struct rb_control_frame_struct *reg_cfp,
+        struct rb_call_info *ci,
+        struct rb_call_cache *cc,
+        VALUE recv))
+{
+    VALUE val;
+    int argc = ci->orig_argc;
+    VALUE recv = TOPN(argc);
+    struct rb_calling_info calling;
+
+    calling.block_handler = block_handler;
+    calling.recv = recv;
+    calling.argc = argc;
+
+    explorer(GET_CFP(), ci, cc, recv);
+
+    val = cc->call(ec, GET_CFP(), &calling, ci, cc);
+
+    if (val != Qundef) {
+        return val;             /* CFUNC normal return */
+    }
+    else {
+        RESTORE_REGS();         /* CFP pushed in cc->cal() */
```

Comment:
@k0kubun `RESTORE_REGS()` is needed here because we do `GET_ISEQ()` and `GET_EP()` below.  If we don't restore `reg_cfp` these macro would look at older control frames, which should break things.

Reply:
That makes sense, thank you. As long as it has no significant impact for performance, it should be fine.

Reply:
Another thing in the same line:

> /* CFP pushed in cc->cal() */

Perhaps you meant `cc->call()`?